lib/oauth/multi.go

```diff
@@ -0,0 +1,153 @@
+package oauth
+
+import (
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/khttp"
+	"github.com/enfabrica/enkit/lib/khttp/kcookie"
+	"github.com/enfabrica/enkit/lib/multierror"
+	"github.com/enfabrica/enkit/lib/token"
+	"log"
+	"math/rand"
+	"net/http"
+	"reflect"
+)
+
+type MultiOauth struct {
+	RequiredAuth *Authenticator
+	OptAuth      []*Authenticator
+	Enc          []*token.TypeEncoder
+}
+
+type MultiOAuthState struct {
+	CurrentFlow   int
+	OptIdentities []Identity
+	Extra         interface{}
+}
+
+func (mo *MultiOauth) NewState(Extra interface{}) *MultiOAuthState {
+	return &MultiOAuthState{
+		Extra: Extra,
+	}
+}
+
+var errStateNotType = errors.New("state did not match")
+
+func (mo *MultiOauth) decodeRaw(state interface{}) (*MultiOAuthState, error) {
```

Comment:
see comment below, change to ever only take a string only as parameter, after the change below decodeRaw is only used in one place.

Reply:
changed to decodeString